Enable SMP on intel_s1000 and up_squared_adsp

dcpleung opened the pull request:
This adds the necessary bits to enable SMP to work on `intel_s1000` and `intel_apl_adsp` SoCs and `intel_s1000_crb` and `up_squared_adsp` board.

- Adds a new entry point for core other than #0. This is due to the fact that Xtensa reset vector for all cores is the same. So only core #0 need to go through the kernel startup routines. The other cores can jump to the function pointed to by `arch_start_cpu()` once basic core initialization is done.
- Adds a sys init level for SMP where the functions are called after `z_smp_init()` where all cores are up and running.
- Adds a driver for the CAVS wall clock timer which is being used as the timer under SMP. This wall clock is external to the CPU cores where it is used as a synchronized timer for all cores.
- Adds a driver to use the CAVS Intra-DSP Communication (IDC) channel as IPI for scheduler.

This passes the SMP test (`tests/kernel/smp`).

dcpleung committed:
interrupt_controller/dw_ictl: need to be initialized earlier

The dw_ictl interrupt controller is an interrupt aggregator
supporting multi-level interrupts. Therefore, it needs to be
initialized earlier than any downstream interrupt controllers
and devices.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
arch/xtensa: smp: only zero BSS only when boot from CPU #0

Under SMP, the main BSS section only needs to be zero-ed on CPU #0.
Other CPUs should not zero out BSS, or else it may cause CPU #0 to
crash on invalid data.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
interrupt_controller/cavs_ictl: need to be initialized earlier

The cavs_ictl interrupt controller is an interrupt aggregator
supporting multi-level interrupts. Therefore, it needs to be
initialized earlier than any downstream interrupt controllers
and devices.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
soc: intel_s1000: use CAVS DSP wall clock timer for SMP

The DSP wall clock timer is a timer driven directly by
external oscillator and is external to the CPU core(s).
It provides a common and synchronized counter for all
CPU cores (which is useful for SMP), instead of
indepedently running local core timer (xtensa_timer).

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
soc: intel_s1000: use EXCSAVE7 for CPU pointer

EXCSAVE2 is used for level 2 exception save location.
Since we are using level 2 interrupts, use EXCSAVE7
instead as level 7 interrupts are not being used
by the SoC.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
soc: intel_s1000: add SMP support

This adds SMP support for Intel S1000 SoC.

Some of the start-up code is borrowed from ESP32.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
xtensa: add calling entry point for multi-processing

Under multi-processing, only the first CPU#0 needs to go through
setting up the kernel structs and clearing out BSS (among others).
There is no need for other CPUs to do those tasks. Since each
Xtensa core starts using the same boot vector, CPUs other than #0
need to skip all the startup tasks by not calling to z_cstart().
So provide another entry point for those CPUs. Note that Xtensa
arch is highly configurable. So the implementation of the entry
point is up to each individual SoC config.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
timer: add CAVS DSP wall clock timer for Intel SoC

The DSP wall clock timer on some Intel SoC is a timer driven
directly by external oscillator and is external to the CPU
core(s). It is not as fast as the internal core clock, but
provides a common and synchronized counter for all CPU cores
(which is useful for SMP).

This uses the RISCV timer as base as it is using 64-bit
counter.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
soc: intel_apl_adsp: select CONFIG_ATOMIC_OPERATIONS_BUILTIN

The GCC 9.2 toolchain for intel_apl_adsp has builtin atomic
functions already so we can use that instead of the custom
one in arch. This allows the atomic function calls to be
inlined.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
dts: ipm: add binding for intel,cavs-idc

This adds the DTS binding for the CAVS Intra-DSP Communication (IDC)
driver for the DSP on Intel SoC.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
kernel: add sys init level for SMP

This adds a sys init level which allows device and sys_init
to be done after SMP initialization, z_smp_init(), when all
cores are up and running.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
interrupt_controller: cavs: add support in ISR for SMP

The CAVS interrupt controller has different base addresses for
each CPU. When running under SMP, the driver needs to look at
the correct address for the CPU the ISR is running so interrupts
can be dispatched correctly. This adds a function to calculate
the correct base address. Note that each supported SoC may have
different offsets so per SoC config will need to added. Support
for intel_s1000 is added as an example.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
interrupt_controller: cavs: add CPU offset for intel_apl_adsp

This adds the per CPU address offset for intel_apl_adsp, so
the correct base address can be calculated under SMP.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

andrewboie commented on include/init.h:
```diff
@@ -25,6 +25,10 @@ extern "C" {
 #define _SYS_INIT_LEVEL_APPLICATION	3
 
+#ifdef CONFIG_SMP
+#define _SYS_INIT_LEVEL_SMP		4
```
I'm not sure I understand why we need a dedicated init level, wouldn't getting all this up and running at POST_KERNEL be sufficient?

dcpleung replied:
POST_KERNEL is done before `z_smp_init()` is called (in `bg_thread_main()`). For something that must be done after all cores are up and running, we need another init level.

andrewboie replied:
I see, wonder if we should just move z_smp_init() to z_cstart()?

andrewboie commented on kernel/sched.c:
```diff
@@ -1143,28 +1140,14 @@ void z_impl_k_wakeup(k_tid_t thread)
-			_current_cpu->swap_ok = true;
-		}
-	}
+	/* NOTE: When adding code to this, make sure this is called
```
since this function is now a no-op, and scoped private with z_, can we just delete it?

dcpleung replied:
`z_sched_ipi()` is the ISR for IPI. We can possibly delete it but I will let @andyross decide.

dcpleung committed:
ipm: add driver for the CAVS DSP Intra-DSP Communication (IDC)

This adds a rather primitive driver for use with the Intra-DSP
Communication (IDC) on the DSP on certain Intel SoCs. The IDC
generates interrupts from one core to another by writing to
certain registers. This is also being utilized as
the scheduler IPI since it can interrupt other cores.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
soc: intel_apl_adsp: add multi-processing support

This adds multi-processing support for Intel Apollo Lake ADSP.

Some of the start-up code is borrowed from ESP32.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

andyross commented on arch/xtensa/core/crt1.S:
```diff
@@ -136,6 +137,14 @@ _start:
+	/* Only clear BSS when running on core 0 */
+	rsr	a3, PRID
+	extui	a3, a3, 0, 8	/* extract core ID */
+	bnez	a3, .L3zte
```
Obligatory whine that stuff like this, which can be implemented as a memset(), really ought to be in C code.  But not an issue for this patch obviously.

andyross commented on drivers/timer/cavs_timer.c:
```diff
@@ -0,0 +1,179 @@
+	u64_t next = last_count + CYC_PER_TICK;
+
+	if ((s64_t)(next - curr) < MIN_DELAY) {
+		next += CYC_PER_TICK;
```
This breaks if MIN_DELAY is larger than one tick.  Not a problem given the numbers above, I think.  Might be worth a BUILD_ASSERT just to be sure.

dcpleung replied:
Added a build assert in the driver.

andyross commented on drivers/timer/cavs_timer.c:
```diff
@@ -0,0 +1,179 @@
+	 * this interrupt is likely the same interrupt
+	 * event but being processed by another CPU.
+	 * Since it has already been processed and
+	 * ticks announced, skip it.
```
The interrupt enable code seems to be putting this one just one CPU.  Is this actually needed?

dcpleung replied:
This is simply to limit the processing to the first CPU that gets the interrupt (note that it's not the first logical CPU, but chronologically first in terms of each timer interrupt). It is an attempt to prevent other cores from spinning and then submit a bunch of unnecessary zero tick announcements.

andyross reviewed (approved):
Nitpicks
